In [1]:
import sys
sys.path.insert(0, '../../')
from panel_exp.panel_data import long_df_to_paneldataset, PanelDataset, TimePeriod

In [2]:
import pandas as pd 
import numpy as np

long_df = pd.read_csv('../data/kansas_parsed.csv', index_col=0)
long_df.iloc[:,0] = long_df.iloc[:,0].apply(lambda x: np.exp(x))
long_df.head()

FileNotFoundError: [Errno 2] No such file or directory: '../data/kansas_parsed.csv'

In [3]:
from panel_exp.design import CompleteRandomization, ThinningDesign, Rerandomization, QuickBlock, greedy_match_markets
from panel_exp.pretest_analysis.mde_ci_module import mde_ci_matrix

In [11]:
panel_data = long_df_to_paneldataset(long_df, "year_qtr", "state", "lngdp")

gm = greedy_match_markets()
td = ThinningDesign(treatment_probability=0.5)
cr = CompleteRandomization(treatment_probability=0.5)
# rr= Rerandomization(treatment_probability=0.5, base_randomizer_cls=greedy_match_markets())

design = [gm,td,cr][0]

test_whitelist = ['Utah']

#to make test whitelist strictly inclusive, you must pass test_blacklist in the below manner (i.e. include all dmas not part of test_whitelist)
# test_blacklist = [geo for geo in panel_data.units.tolist() if geo not in test_whitelist]

gm_panel_data = design.assign(panel_data=panel_data,test_whitelist=test_whitelist)

running iteration 1
running iteration 2
running iteration 3
running iteration 4
running iteration 5
running iteration 6
running iteration 7
running iteration 8
running iteration 9
running iteration 10
running iteration 11
running iteration 12
running iteration 13
running iteration 14
running iteration 15
running iteration 16
running iteration 17
running iteration 18
running iteration 19
running iteration 20


/Users/ppavuluri/Desktop/panel_exp/panel_exp/examples/../panel_exp/panel_data.py:515: FutureWarning: In a future version, the Index constructor will not infer numeric dtypes when passed object-dtype sequences (matching Series behavior)
  df = df.pivot(columns=time_column, values=value_column)


In [12]:
print('Test group: %s \n\nControl group: %s' %(gm_panel_data.treated_units, [unit for unit in gm_panel_data.units if unit not in gm_panel_data.treated_units]))


Test group: ['Utah', 'Vermont', 'New Jersey', 'Kansas', 'New Mexico', 'South Carolina', 'Iowa', 'Mississippi', 'South Dakota', 'New Hampshire', 'Kentucky', 'Rhode Island'] 

Control group: ['Alabama', 'Alaska', 'Arkansas', 'Connecticut', 'Delaware', 'Hawaii', 'Idaho', 'Illinois', 'Maine', 'Missouri', 'Montana', 'North Carolina', 'Pennsylvania', 'Wisconsin', 'Wyoming']


Below is the sensitivity analysis for test and control groups identified by the selected design. Additionally, we also provide sensitivity analysis for trimmed versions of test group incase we need to reduce the test size due to spend constraints.

In [ ]:
results = mde_ci_matrix(gm_panel_data)

In [15]:
#Minimum detectable effect for varying test sizes
results['mde_scores']

,16wk_percent,12wk_percent,8wk_percent,6wk_percent,4wk_percent,test_sales_prc,control_sales_prc,test_dmas,control_dmas,test_dmas_len,control_dmas_len
0,0.141818,0.138477,0.152077,0.153609,0.230806,34,66,"[South Dakota, New Hampshire, Kentucky, New Je...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",12,15
1,0.360816,0.326691,0.374737,0.412488,0.489044,23,66,"[South Dakota, New Hampshire, South Carolina, ...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",11,15
2,0.260035,0.262633,0.300891,0.369090,0.349095,20,66,"[South Dakota, New Hampshire, Kentucky, South ...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",10,15
3,0.251403,0.244189,0.266657,0.301268,0.360467,16,66,"[South Dakota, New Hampshire, Kentucky, South ...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",8,15
4,0.252369,0.246200,0.277439,0.335163,0.345515,13,66,"[South Dakota, New Hampshire, South Carolina, ...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",7,15
5,0.286338,0.350505,0.352827,0.420887,0.401865,10,66,"[South Dakota, New Hampshire, Rhode Island, Mi...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",6,15
6,0.379026,0.390738,0.416088,0.480061,0.481896,6,66,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",5,15
7,0.507056,0.590508,0.673805,0.640748,0.716624,4,66,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",4,15
8,0.766733,0.801776,0.898062,0.930422,1.041409,3,66,"[New Hampshire, Rhode Island, Vermont]","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",3,15
9,0.631823,0.708808,0.742419,0.709514,0.949946,2,66,"[New Hampshire, Vermont]","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",2,15


In [16]:
#Confidence Interval for varying test sizes
results['ci_scores']

,16wk_abs,12wk_abs,8wk_abs,6wk_abs,4wk_abs,test_sales_prc,control_sales_prc,test_dmas,control_dmas,test_dmas_len,control_dmas_len
0,190519.0,139161.0,102425.0,84196.0,71961.0,34,66,"[South Dakota, New Hampshire, Kentucky, New Je...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",12,15
1,313565.0,232298.0,179530.0,142900.0,108241.0,23,66,"[South Dakota, New Hampshire, South Carolina, ...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",11,15
2,208945.0,160874.0,113178.0,103151.0,71759.0,20,66,"[South Dakota, New Hampshire, Kentucky, South ...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",10,15
3,157816.0,125966.0,83638.0,70684.0,54753.0,16,66,"[South Dakota, New Hampshire, Kentucky, South ...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",8,15
4,133471.0,100232.0,75358.0,70987.0,46642.0,13,66,"[South Dakota, New Hampshire, South Carolina, ...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",7,15
5,110524.0,102107.0,67760.0,56552.0,43409.0,10,66,"[South Dakota, New Hampshire, Rhode Island, Mi...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",6,15
6,93194.0,68029.0,48924.0,43115.0,29719.0,6,66,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",5,15
7,78027.0,68441.0,51092.0,36172.0,27339.0,4,66,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",4,15
8,91704.0,73142.0,54726.0,40549.0,30851.0,3,66,"[New Hampshire, Rhode Island, Vermont]","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",3,15
9,52623.0,39985.0,28582.0,21819.0,17534.0,2,66,"[New Hampshire, Vermont]","[Alabama, Alaska, Arkansas, Connecticut, Delaw...",2,15


Trim control group for a given test group - if we were to choose trimmed version of test group from the results above for spend reasons, then we want to ensure control group is also updated accordingly - i.e. check if we can trim down control without compromising on sensitivity. This way we can run the test with smaller control and save those trimmed geos for another test. 

As an example, I'm picking trimmed test group from row 6 where 8 week MDE and CI are 0.41%, 48.9K respectively, then apply control trim to see if it's possible to further improve sensitivity by trimming down the control group. 

In [24]:
#Limiting the raw data to just control and test geos
trimmed_long_df = long_df[long_df.state.isin(results['ci_scores'].iloc[6,7] + results['ci_scores'].iloc[6,8])]

In [25]:
#setting treated units
treated_units = results['ci_scores'].iloc[6,7]

In [26]:
#creating panel dataset with treated units
trimmed_panel_data = long_df_to_paneldataset(trimmed_long_df, "year_qtr", "state", "lngdp", treated_units = treated_units)

In [ ]:
trimmed_results = mde_ci_matrix(trimmed_panel_data, test_trim='no', control_trim='yes')

In [28]:
trimmed_results['mde_scores']

,16wk_percent,12wk_percent,8wk_percent,6wk_percent,4wk_percent,test_sales_prc,control_sales_prc,test_dmas,control_dmas,test_dmas_len,control_dmas_len
0,0.371486,0.423437,0.407092,0.481119,0.548864,8,92,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Alaska, Pennsylvania, Delaware, Ill...",5,15
1,0.347463,0.344662,0.364451,0.449895,0.522619,8,74,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Alaska, Illinois, Delaware, Maine, ...",5,14
2,0.311335,0.300597,0.363352,0.434242,0.463758,8,54,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Alaska, Delaware, Maine, Idaho, Mis...",5,13
3,0.311696,0.320195,0.357167,0.373099,0.506448,8,41,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Alaska, Delaware, Maine, Idaho, Mis...",5,12
4,0.283579,0.295324,0.325836,0.342643,0.380974,8,34,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Alaska, Delaware, Maine, Idaho, Mis...",5,11
5,0.296341,0.315628,0.335027,0.384465,0.453811,8,25,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Delaware, Maine, Missouri, Idaho, H...",5,9
6,0.457863,0.542389,0.494768,0.598585,0.592260,8,17,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Delaware, Maine, Idaho, Hawaii, Mon...",5,8
7,0.328519,0.356565,0.381128,0.432734,0.494770,8,15,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Delaware, Maine, Idaho, Montana, Al...",5,7
8,0.333686,0.343849,0.399192,0.408809,0.435421,8,13,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Maine, Idaho, Montana, Alabama]",5,5
9,0.369227,0.369866,0.437318,0.473522,0.485895,8,7,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Montana, Maine, Arkansas, Idaho]",5,4


In [29]:
trimmed_results['ci_scores']

,16wk_abs,12wk_abs,8wk_abs,6wk_abs,4wk_abs,test_sales_prc,control_sales_prc,test_dmas,control_dmas,test_dmas_len,control_dmas_len
0,84110.0,72792.0,49750.0,41159.0,32519.0,8,92,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Alaska, Pennsylvania, Delaware, Ill...",5,15
1,80517.0,61838.0,45212.0,41626.0,31405.0,8,74,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Alaska, Illinois, Delaware, Maine, ...",5,14
2,74227.0,54573.0,42271.0,35981.0,27719.0,8,54,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Alaska, Delaware, Maine, Idaho, Mis...",5,13
3,74641.0,59449.0,43395.0,32898.0,30068.0,8,41,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Alaska, Delaware, Maine, Idaho, Mis...",5,12
4,67602.0,51839.0,38045.0,31801.0,22723.0,8,34,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Alaska, Delaware, Maine, Idaho, Mis...",5,11
5,69720.0,55114.0,40517.0,34879.0,26658.0,8,25,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Delaware, Maine, Missouri, Idaho, H...",5,9
6,111437.0,94340.0,58483.0,51919.0,39257.0,8,17,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Delaware, Maine, Idaho, Hawaii, Mon...",5,8
7,77551.0,60281.0,45990.0,37189.0,29304.0,8,15,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Delaware, Maine, Idaho, Montana, Al...",5,7
8,79789.0,65495.0,45614.0,36579.0,27860.0,8,13,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Arkansas, Maine, Idaho, Montana, Alabama]",5,5
9,86715.0,67123.0,52862.0,41861.0,27969.0,8,7,"[South Dakota, New Hampshire, Rhode Island, Ve...","[Montana, Maine, Arkansas, Idaho]",5,4


As you can see from row 4 in the above results that 8 week MDE and CI (0.32%, 38K) significantly improved after trimming down control from 15 geos to 11 geos.

To recap, if we are using non-trimmed version of the test group then no further trimming of control is needed but if we decide to use trimmed version of test group due to spend constraints then we'd want to check if applying control trim yeilds better sensitivity.